src/CodeGen/ODataT4CodeGenTemplate.ttinclude

```diff
@@ -0,0 +1,1333 @@
+                    }
+
+                    string fixedActionName = GetFixedName(actionName);
+                    string ac = string.Format("{0}({1},{2})", fixedActionName, sourceTypeName, parameterTypes );
+                    if (!boundOperations.Contains(ac))
+                    {
+                        boundOperations.Add(ac);
+                        this.WriteBoundActionAsExtension(fixedActionName, action.Name, sourceTypeName, returnTypeName, parameterString, action.Namespace, parameterValues);
+                    }
+
+                    IEdmStructuredType structuredType;
+                    if (edmTypeReference.IsCollection())
+                    {
+                        IEdmCollectionType collectionType = edmTypeReference.Definition as IEdmCollectionType;
+                        structuredType = (IEdmStructuredType)collectionType.ElementType.Definition;
+                    }
+                    else
+                    {
+                        structuredType = (IEdmStructuredType)edmTypeReference.Definition;
+                    }
+
+                    List<IEdmStructuredType> derivedTypes;
+                    if (structuredBaseTypeMap.TryGetValue(structuredType, out derivedTypes))
+                    {
+                        foreach (IEdmStructuredType type in derivedTypes)
+                        {
+                            IEdmTypeReference derivedTypeReference = new EdmEntityTypeReference((IEdmEntityType)type, true);
+                            List<IEdmTypeReference> currentParameters = action.Parameters.Select(p => p.Type).ToList();
+                            currentParameters[0] = derivedTypeReference;
+
+                            sourceTypeName = string.Format(edmTypeReference.IsCollection() ? this.DataServiceQueryStructureTemplate : this.DataServiceQuerySingleStructureTemplate, GetSourceOrReturnTypeName(derivedTypeReference));
+                            string currentAc = string.Format("{0}({1},{2})", fixedActionName, sourceTypeName, parameterTypes );
+                            if (!boundOperations.Contains(currentAc))
+                            {
+                                boundOperations.Add(currentAc);
+                                this.WriteBoundActionAsExtension(fixedActionName, action.Name, sourceTypeName, returnTypeName, parameterString, action.Namespace, parameterValues);
+                            }
+                        }
+                    }
+                }
+            }
+
+            this.WriteExtensionMethodsEnd();
+        }
+
+        this.WriteNamespaceEnd();
+    }
+    
+    internal bool HasBoundOperations(IEnumerable<IEdmOperation> operations)
+    {
+        foreach (IEdmOperation opeartion in operations)
+        {
+            if (opeartion.IsBound)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    internal void WriteEntityContainer(IEdmEntityContainer container, string fullNamespace)

```

Comment:
This function can also be broken down into smaller functions.

Reply:
Fixed